### PR TITLE
executor,tablecodec: optimize decodeIndexKeyValue function for memIndexReader

### PR DESCRIPTION
**Pull request by tiancaiamao**
<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
<!--

Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.

-->

Issue Number: ref #43249 

Problem Summary:

Too many allocation for the decode operation.

### What changed and how does it work?

Reuse memory to avoid allocations.

### Check List

Tests <!-- At least one of them must be included. -->

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)


```
mysql> show create table AccessoryMaster;
+-----------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table           | Create Table                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
+-----------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| AccessoryMaster | CREATE TABLE `AccessoryMaster` (
  `id` int(11) unsigned NOT NULL,
  `name` varchar(45) NOT NULL COMMENT '名前',
  `nameId` int(11) unsigned NOT NULL,
  `description` varchar(256) DEFAULT NULL COMMENT '効果説明',
  `rarity` tinyint(3) unsigned NOT NULL COMMENT 'レアリティ',
  `weight` tinyint(3) unsigned NOT NULL DEFAULT '1' COMMENT '重み',
  `accessorySkillType1` tinyint(3) NOT NULL DEFAULT '0' COMMENT '0:なし、1:アクションスキル、2:パッシブスキル',
  `accessorySkillId1` int(10) NOT NULL DEFAULT '0' COMMENT 'スキルID1（0:無し）',
  `accessorySkillType2` tinyint(3) NOT NULL DEFAULT '0' COMMENT '0:なし、1:アクションスキル、2:パッシブスキル',
  `accessorySkillId2` int(10) NOT NULL DEFAULT '0' COMMENT 'スキルID2（0:無し）',
  `accessorySkillType3` tinyint(3) NOT NULL DEFAULT '0' COMMENT '0:なし、1:アクションスキル、2:パッシブスキル',
  `accessorySkillId3` int(10) NOT NULL DEFAULT '0' COMMENT 'スキルID3（0:無し）',
  `accessorySkillType4` tinyint(3) NOT NULL DEFAULT '0' COMMENT '0:なし、1:アクションスキル、2:パッシブスキル',
  `accessorySkillId4` int(10) NOT NULL DEFAULT '0' COMMENT 'スキルID4（0:無し）',
  `sellMoney` smallint(11) NOT NULL COMMENT '売価ゴールド',
  `noDuplicateFlag` tinyint(3) unsigned NOT NULL DEFAULT '0' COMMENT '重複不可フラグ',
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,
  KEY `nameId` (`nameId`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin COMMENT='アクセサリーマスタ' /* CACHED ON */ |
+-----------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)


mysql>  select count(*) from  AccessoryMaster;
+----------+
| count(*) |
+----------+
|   292767 |
+----------+
1 row in set (1.15 sec)
```

Benchmark with `SELECT id FROM AccessoryMaster` and profiling.


Before:

![image](https://github.com/user-attachments/assets/310881d8-ad29-4368-8e28-b16dd1ad39f0)

After:

![image](https://github.com/user-attachments/assets/f2a7b768-279b-42ae-a5d0-3c0ea4b4ce16)



- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > <!-- Or your custom  "No need to test" reasons -->

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/executor/mem_reader.go [hunk 1]
```diff
@@ -199,7 +212,10 @@ func (m *memIndexReader) decodeIndexKeyValue(key, value []byte, tps []*types.Fie
 		physTblIDColumnIdx = m.outputOffset[m.physTblIDIdx]
 	}
 
-	ds := make([]types.Datum, 0, len(m.outputOffset))
+	if m.resultRows == nil {
+		m.resultRows = make([]types.Datum, 0, len(m.outputOffset))
+	}
```
Comment (Defined2014) [inline pkg/executor/mem_reader.go:217]: Maybe we can init it when create `memIndexReader`

**Review by Defined2014 — APPROVED** [review]
Others LGTM

File: pkg/executor/mem_reader.go (hunk 1 above)
Comment (tiancaiamao) [reply pkg/executor/mem_reader.go]: done

**Review by YangKeao — APPROVED** [review]
LGTM

**Review by zyguan — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**